// framer/build/router/stub.js
/* eslint-disable no-console */
/**
 * @module
 * This module provides stubs for the Router API used in Handshake instead of the private Framer Router.
 * */
import * as React from "react";
export const ComponentWithPreload = stubbedComponent("ComponentWithPreload");
export const RoutesProvider = stubbedComponent("RoutesProvider");
export const Router = stubbedComponent("Router");
export function computeRelativePath(_from, _to) {
    return "";
}
export function isRoute(route) {
    return false;
}
export function getRouteElementId(_route, _elementId) {
    return undefined;
}
export function useCurrentRouteId() {
    return undefined;
}
export function useRouter() {
    return {
        navigate(_routeId, _elementId) { },
        getRoute(_routeId, _elementId) {
            return undefined;
        },
        currentRouteId: "",
    };
}
/**
 * @internal
 */
export function useRouteAnchor(_routeId, { elementId } = {}) {
    return { onClick: () => { }, href: "#" };
}
/**
 * @internal
 */
export function useRouteHandler(_routeId, _preload = false, elementId) {
    return () => { };
}
export function useRoutePreloader(_routeIds, _enabled = true) {
    return;
}
function stubbedComponent(componentName) {
    return ({ children }) => {
        console.warn(`Using stubbed \`${componentName}\``);
        return React.createElement(React.Fragment, null, children);
    };
}
