// @parcel/core/lib/public/BundleGraph.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.bundleGraphToInternalBundleGraph = bundleGraphToInternalBundleGraph;
exports.default = void 0;

function _assert() {
  const data = _interopRequireDefault(require("assert"));

  _assert = function () {
    return data;
  };

  return data;
}

function _nullthrows() {
  const data = _interopRequireDefault(require("nullthrows"));

  _nullthrows = function () {
    return data;
  };

  return data;
}

function _Asset() {
  const data = require("./Asset");

  _Asset = function () {
    return data;
  };

  return data;
}

function _Bundle() {
  const data = require("./Bundle");

  _Bundle = function () {
    return data;
  };

  return data;
}

function _Dependency() {
  const data = _interopRequireWildcard(require("./Dependency"));

  _Dependency = function () {
    return data;
  };

  return data;
}

function _Graph() {
  const data = require("../Graph");

  _Graph = function () {
    return data;
  };

  return data;
}

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classPrivateFieldGet(receiver, privateMap) { var descriptor = _classExtractFieldDescriptor(receiver, privateMap, "get"); return _classApplyDescriptorGet(receiver, descriptor); }

function _classApplyDescriptorGet(receiver, descriptor) { if (descriptor.get) { return descriptor.get.call(receiver); } return descriptor.value; }

function _classPrivateFieldSet(receiver, privateMap, value) { var descriptor = _classExtractFieldDescriptor(receiver, privateMap, "set"); _classApplyDescriptorSet(receiver, descriptor, value); return value; }

function _classExtractFieldDescriptor(receiver, privateMap, action) { if (!privateMap.has(receiver)) { throw new TypeError("attempted to " + action + " private field on non-instance"); } return privateMap.get(receiver); }

function _classApplyDescriptorSet(receiver, descriptor, value) { if (descriptor.set) { descriptor.set.call(receiver, value); } else { if (!descriptor.writable) { throw new TypeError("attempted to set read only private field"); } descriptor.value = value; } }

// Friendly access for other modules within this package that need access
// to the internal bundle.
const _bundleGraphToInternalBundleGraph = new WeakMap();

function bundleGraphToInternalBundleGraph(bundleGraph) {
  return (0, _nullthrows().default)(_bundleGraphToInternalBundleGraph.get(bundleGraph));
}

var _graph = new WeakMap();

var _options = new WeakMap();

var _createBundle = new WeakMap();

class BundleGraph {
  // This is invoked as `this.#createBundle.call(null, ...)` below, as private
  // properties aren't currently callable in Flow:
  // https://github.com/parcel-bundler/parcel/pull/4591#discussion_r422661115
  // https://github.com/facebook/flow/issues/7877
  constructor(graph, createBundle, options) {
    _graph.set(this, {
      writable: true,
      value: void 0
    });

    _options.set(this, {
      writable: true,
      value: void 0
    });

    _createBundle.set(this, {
      writable: true,
      value: void 0
    });

    _classPrivateFieldSet(this, _graph, graph);

    _classPrivateFieldSet(this, _options, options);

    _classPrivateFieldSet(this, _createBundle, createBundle); // $FlowFixMe


    _bundleGraphToInternalBundleGraph.set(this, graph);
  }

  getAssetById(id) {
    return (0, _Asset().assetFromValue)(_classPrivateFieldGet(this, _graph).getAssetById(id), _classPrivateFieldGet(this, _options));
  }

  getAssetPublicId(asset) {
    return _classPrivateFieldGet(this, _graph).getAssetPublicId((0, _Asset().assetToAssetValue)(asset));
  }

  isDependencySkipped(dep) {
    return _classPrivateFieldGet(this, _graph).isDependencySkipped((0, _Dependency().dependencyToInternalDependency)(dep));
  }

  getDependencyResolution(dep, bundle) {
    let resolution = _classPrivateFieldGet(this, _graph).getDependencyResolution((0, _Dependency().dependencyToInternalDependency)(dep), bundle && (0, _Bundle().bundleToInternalBundle)(bundle));

    if (resolution) {
      return (0, _Asset().assetFromValue)(resolution, _classPrivateFieldGet(this, _options));
    }
  }

  getIncomingDependencies(asset) {
    return _classPrivateFieldGet(this, _graph).getIncomingDependencies((0, _Asset().assetToAssetValue)(asset)).map(dep => new (_Dependency().default)(dep));
  }

  getAssetWithDependency(dep) {
    let asset = _classPrivateFieldGet(this, _graph).getAssetWithDependency((0, _Dependency().dependencyToInternalDependency)(dep));

    if (asset) {
      return (0, _Asset().assetFromValue)(asset, _classPrivateFieldGet(this, _options));
    }
  }

  getBundleGroupsContainingBundle(bundle) {
    return _classPrivateFieldGet(this, _graph).getBundleGroupsContainingBundle((0, _Bundle().bundleToInternalBundle)(bundle));
  }

  getReferencedBundles(bundle, opts) {
    return _classPrivateFieldGet(this, _graph).getReferencedBundles((0, _Bundle().bundleToInternalBundle)(bundle), opts).map(bundle => _classPrivateFieldGet(this, _createBundle).call(null, bundle, _classPrivateFieldGet(this, _graph), _classPrivateFieldGet(this, _options)));
  }

  getRequiredBundlesForBundle(bundle) {
    return _classPrivateFieldGet(this, _graph).getReferencedBundles((0, _Bundle().bundleToInternalBundle)(bundle)).map(bundle => _classPrivateFieldGet(this, _createBundle).call(null, bundle, _classPrivateFieldGet(this, _graph), _classPrivateFieldGet(this, _options)));
  }

  resolveAsyncDependency(dependency, bundle) {
    let resolved = _classPrivateFieldGet(this, _graph).resolveAsyncDependency((0, _Dependency().dependencyToInternalDependency)(dependency), bundle && (0, _Bundle().bundleToInternalBundle)(bundle));

    if (resolved == null) {
      return;
    } else if (resolved.type === 'bundle_group') {
      return resolved;
    }

    return {
      type: 'asset',
      value: (0, _Asset().assetFromValue)(resolved.value, _classPrivateFieldGet(this, _options))
    };
  }

  getReferencedBundle(dependency, bundle) {
    let result = _classPrivateFieldGet(this, _graph).getReferencedBundle((0, _Dependency().dependencyToInternalDependency)(dependency), (0, _Bundle().bundleToInternalBundle)(bundle));

    if (result != null) {
      return _classPrivateFieldGet(this, _createBundle).call(null, result, _classPrivateFieldGet(this, _graph), _classPrivateFieldGet(this, _options));
    }
  }

  getDependencies(asset) {
    return _classPrivateFieldGet(this, _graph).getDependencies((0, _Asset().assetToAssetValue)(asset)).map(dep => new (_Dependency().default)(dep));
  }

  isAssetReachableFromBundle(asset, bundle) {
    return _classPrivateFieldGet(this, _graph).isAssetReachableFromBundle((0, _Asset().assetToAssetValue)(asset), (0, _Bundle().bundleToInternalBundle)(bundle));
  }

  findReachableBundleWithAsset(bundle, asset) {
    let result = _classPrivateFieldGet(this, _graph).findReachableBundleWithAsset((0, _Bundle().bundleToInternalBundle)(bundle), (0, _Asset().assetToAssetValue)(asset));

    if (result != null) {
      return _classPrivateFieldGet(this, _createBundle).call(null, result, _classPrivateFieldGet(this, _graph), _classPrivateFieldGet(this, _options));
    }

    return null;
  }

  isAssetReferencedByDependant(bundle, asset) {
    return _classPrivateFieldGet(this, _graph).isAssetReferencedByDependant((0, _Bundle().bundleToInternalBundle)(bundle), (0, _Asset().assetToAssetValue)(asset));
  }

  hasParentBundleOfType(bundle, type) {
    return _classPrivateFieldGet(this, _graph).hasParentBundleOfType((0, _Bundle().bundleToInternalBundle)(bundle), type);
  }

  getBundlesInBundleGroup(bundleGroup) {
    return _classPrivateFieldGet(this, _graph).getBundlesInBundleGroup(bundleGroup).map(bundle => _classPrivateFieldGet(this, _createBundle).call(null, bundle, _classPrivateFieldGet(this, _graph), _classPrivateFieldGet(this, _options)));
  }

  getBundles() {
    return _classPrivateFieldGet(this, _graph).getBundles().map(bundle => _classPrivateFieldGet(this, _createBundle).call(null, bundle, _classPrivateFieldGet(this, _graph), _classPrivateFieldGet(this, _options)));
  }

  isEntryBundleGroup(bundleGroup) {
    return _classPrivateFieldGet(this, _graph).isEntryBundleGroup(bundleGroup);
  }

  getChildBundles(bundle) {
    return _classPrivateFieldGet(this, _graph).getChildBundles((0, _Bundle().bundleToInternalBundle)(bundle)).map(bundle => _classPrivateFieldGet(this, _createBundle).call(null, bundle, _classPrivateFieldGet(this, _graph), _classPrivateFieldGet(this, _options)));
  }

  getParentBundles(bundle) {
    return _classPrivateFieldGet(this, _graph).getParentBundles((0, _Bundle().bundleToInternalBundle)(bundle)).map(bundle => _classPrivateFieldGet(this, _createBundle).call(null, bundle, _classPrivateFieldGet(this, _graph), _classPrivateFieldGet(this, _options)));
  }

  resolveSymbol(asset, symbol, boundary) {
    let res = _classPrivateFieldGet(this, _graph).resolveSymbol((0, _Asset().assetToAssetValue)(asset), symbol, boundary ? (0, _Bundle().bundleToInternalBundle)(boundary) : null);

    return {
      asset: (0, _Asset().assetFromValue)(res.asset, _classPrivateFieldGet(this, _options)),
      exportSymbol: res.exportSymbol,
      symbol: res.symbol,
      loc: res.loc
    };
  }

  getExportedSymbols(asset, boundary) {
    let res = _classPrivateFieldGet(this, _graph).getExportedSymbols((0, _Asset().assetToAssetValue)(asset), boundary ? (0, _Bundle().bundleToInternalBundle)(boundary) : null);

    return res.map(e => ({
      asset: (0, _Asset().assetFromValue)(e.asset, _classPrivateFieldGet(this, _options)),
      exportSymbol: e.exportSymbol,
      symbol: e.symbol,
      loc: e.loc,
      exportAs: e.exportAs
    }));
  }

  traverse(visit) {
    return _classPrivateFieldGet(this, _graph).traverse((0, _Graph().mapVisitor)(node => node.type === 'asset' ? {
      type: 'asset',
      value: (0, _Asset().assetFromValue)(node.value, _classPrivateFieldGet(this, _options))
    } : {
      type: 'dependency',
      value: new (_Dependency().default)(node.value)
    }, visit));
  }

  traverseBundles(visit, startBundle) {
    return _classPrivateFieldGet(this, _graph).traverseBundles((0, _Graph().mapVisitor)(bundle => _classPrivateFieldGet(this, _createBundle).call(null, bundle, _classPrivateFieldGet(this, _graph), _classPrivateFieldGet(this, _options)), visit), startBundle == null ? undefined : (0, _Bundle().bundleToInternalBundle)(startBundle));
  }

  findBundlesWithAsset(asset) {
    return _classPrivateFieldGet(this, _graph).findBundlesWithAsset((0, _Asset().assetToAssetValue)(asset)).map(bundle => _classPrivateFieldGet(this, _createBundle).call(null, bundle, _classPrivateFieldGet(this, _graph), _classPrivateFieldGet(this, _options)));
  }

  findBundlesWithDependency(dependency) {
    return _classPrivateFieldGet(this, _graph).findBundlesWithDependency((0, _Dependency().dependencyToInternalDependency)(dependency)).map(bundle => _classPrivateFieldGet(this, _createBundle).call(null, bundle, _classPrivateFieldGet(this, _graph), _classPrivateFieldGet(this, _options)));
  }

  getUsedSymbols(v) {
    if (v instanceof _Asset().Asset) {
      return _classPrivateFieldGet(this, _graph).getUsedSymbolsAsset((0, _Asset().assetToAssetValue)(v));
    } else {
      (0, _assert().default)(v instanceof _Dependency().default);
      return _classPrivateFieldGet(this, _graph).getUsedSymbolsDependency((0, _Dependency().dependencyToInternalDependency)(v));
    }
  }

}

exports.default = BundleGraph;
