// framer/build/render/presentation/RichText.js
import { useRoutePreloader, useRouter } from "../../router/index.js";
import { motion } from "framer-motion";
import * as React from "react";
import { useRef } from "react";
import { Animatable } from "../../animation/Animatable/Animatable.js";
import { cx } from "../../modules/cx.js";
import { parseFramerPageLink, replaceFramerPageLinks, useLinkMatchesRoute, } from "../../modules/framerPageLink.js";
import { useIsOnFramerCanvas } from "../../modules/useIsOnFramerCanvas.js";
import { collectTextShadowsForProps } from "../style/shadow.js";
import { calculateRect, ParentSizeState, useParentSize, } from "../types/NewConstraints.js";
import { RenderTarget } from "../types/RenderEnvironment.js";
import { collectFiltersFromProps } from "../utils/filtersForNode.js";
import { injectComponentCSSRules } from "../utils/injectComponentCSSRules.js";
import { forceLayerBackingWithCSSProperties } from "../utils/setLayerBacked.js";
import { transformTemplate } from "../utils/transformTemplate.js";
import { useLayoutId } from "../utils/useLayoutId.js";
import { useMeasureLayout } from "../utils/useMeasureLayout.js";
import { ComponentContainerContext } from "./ComponentContainerContext.js";
import { findAnchorElementWithin } from "./Text.js";
/**
 * @internal
 * Don't change this value as it's used in generated components as well.
 */
export const richTextPlaceholder = "{{ text-placeholder }}";
/**
 * @internal
 */
export const RichText = React.forwardRef(function Text(props, forwardedRef) {
    const { id, name, html = "", text, width, height, left, right, top, bottom, center, className, visible = true, opacity, rotation = 0, verticalAlignment = "top", isEditable = false, willChangeTransform, environment = RenderTarget.current, withExternalLayout = false, positionSticky, positionStickyTop, positionStickyRight, positionStickyBottom, positionStickyLeft, __htmlStructure, __fromCanvasComponent, _forwardedOverrideId, _forwardedOverrides, _usesDOMRect, ...rest } = props;
    const parentSize = useParentSize();
    const layoutId = useLayoutId(props);
    const fallbackLayoutRef = useRef(null);
    const layoutRef = forwardedRef ?? fallbackLayoutRef;
    const { navigate, getRoute, currentRouteId } = useRouter();
    useRoutePreloader(props.preload ?? []);
    useMeasureLayout(props, layoutRef);
    const inCodeComponent = React.useContext(ComponentContainerContext);
    const matchesCurrentRoute = useLinkMatchesRoute(props.__link);
    const isOnCanvas = useIsOnFramerCanvas();
    let innerHTML = html;
    let textContent = text;
    // Replace HTML with text override or variable
    const forwardedOverrideId = _forwardedOverrideId ?? id;
    if (forwardedOverrideId && _forwardedOverrides) {
        const text = _forwardedOverrides[forwardedOverrideId];
        if (typeof text === "string") {
            textContent = text;
        }
    }
    // The placeholder is inserted in collectRichText.ts
    if (textContent && __htmlStructure) {
        innerHTML = __htmlStructure.replace(richTextPlaceholder, textContent);
    }
    // Parse the ProseMirror HTML to replace framer page links in the `data:framer/page-link,`
    // format with the correct paths their target corresponds to.
    // This needs to happen whenever the HTML string changes, since smart
    // components can update the ProseMirror HTML in variants by updating this prop value.
    innerHTML = React.useMemo(() => {
        if (isOnCanvas || !getRoute || !currentRouteId)
            return innerHTML;
        return replaceFramerPageLinks(innerHTML, getRoute, currentRouteId);
    }, [innerHTML, getRoute, isOnCanvas, currentRouteId]);
    React.useEffect(() => {
        const container = layoutRef.current;
        if (container === null)
            return;
        function interceptPageLinks(event) {
            const anchorElement = findAnchorElementWithin(event.target, layoutRef);
            // Don't prevent cmd + click, allowing users to open in new tab.
            if (event.metaKey || !navigate || !anchorElement)
                return;
            // Page links written to the ProseMirror HTML in the `data:framer/page-link,`
            // format should have been parsed, and replaced with the correct paths.
            // If that has happened, we can access the route id and transition properties
            // as data attributes, finally performing a navigation.
            const pageLink = anchorElement.getAttribute("data-framer-page-link-target" /* Page */);
            const elementId = anchorElement.getAttribute("data-framer-page-link-element" /* Element */);
            if (pageLink) {
                event.preventDefault();
                return navigate(pageLink, elementId ?? undefined);
            }
            // If for some reason the page links in the ProseMirror HTML have not been
            // parsed, try to still perform a navigation by parsing the page
            // link in the href.
            const href = anchorElement.getAttribute("href");
            if (href) {
                const link = parseFramerPageLink(href);
                if (!link || !link.target)
                    return;
                event.preventDefault();
                return navigate(link.target, link.element ?? undefined);
            }
        }
        container.addEventListener("click", interceptPageLinks);
        return () => {
            container.removeEventListener("click", interceptPageLinks);
        };
    }, [navigate]);
    // Hooks need to be above this line
    if (!visible)
        return null;
    injectComponentCSSRules();
    const isHidden = isEditable && environment() === RenderTarget.canvas;
    // Add more styling and support vertical text alignment
    const style = {
        outline: "none",
        display: "flex",
        flexDirection: "column",
        justifyContent: convertVerticalAlignment(verticalAlignment),
        opacity: isHidden ? 0 : opacity,
        flexShrink: 0,
    };
    const restrictedRenderTarget = RenderTarget.hasRestrictions();
    const frame = calculateRect(props, parentSize || ParentSizeState.Unknown, false);
    const isAutoSized = _usesDOMRect && (width === "auto" || height === "auto");
    const hasTransformTemplate = !frame || !restrictedRenderTarget || __fromCanvasComponent || isAutoSized;
    if (!withExternalLayout) {
        if (frame && restrictedRenderTarget && !isAutoSized) {
            const rotate = Animatable.getNumber(rotation).toFixed(4);
            style.transform = `translate(${frame.x}px, ${frame.y}px) rotate(${rotate}deg)`;
            style.width = frame.width;
            style.minWidth = frame.width;
            style.height = frame.height;
        }
        else {
            style.left = left;
            style.right = right;
            style.top = top;
            style.bottom = bottom;
            style.width = width;
            style.height = height;
            // I have no idea why this works as a number but not as a string. This is copied from the Text component.
            style.rotate = rotation;
        }
        if (positionSticky) {
            if (!isOnCanvas || inCodeComponent) {
                style.position = "sticky";
                style.willChange = "transform";
                style.zIndex = 1;
                style.top = positionStickyTop;
                style.right = positionStickyRight;
                style.bottom = positionStickyBottom;
                style.left = positionStickyLeft;
            }
        }
    }
    collectFiltersFromProps(props, style);
    collectTextShadowsForProps(props, style);
    if (willChangeTransform) {
        // We're not using Layer.applyWillChange here, because adding willChange:transform causes clipping issues in export
        forceLayerBackingWithCSSProperties(style);
    }
    Object.assign(style, props.style);
    return (React.createElement(motion.div, { id: id, ref: layoutRef, ...rest, style: style, layoutId: layoutId, "data-framer-name": name, "data-framer-component-type": "RichText", "data-center": center, className: cx(className, matchesCurrentRoute && "current"), transformTemplate: hasTransformTemplate ? transformTemplate(center) : undefined, dangerouslySetInnerHTML: { __html: innerHTML } }));
});
function convertVerticalAlignment(verticalAlignment) {
    switch (verticalAlignment) {
        case "top":
            return "flex-start";
        case "center":
            return "center";
        case "bottom":
            return "flex-end";
    }
}
