// @parcel/utils/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "countLines", {
  enumerable: true,
  get: function () {
    return _countLines().default;
  }
});
Object.defineProperty(exports, "generateBuildMetrics", {
  enumerable: true,
  get: function () {
    return _generateBuildMetrics().default;
  }
});
Object.defineProperty(exports, "generateCertificate", {
  enumerable: true,
  get: function () {
    return _generateCertificate().default;
  }
});
Object.defineProperty(exports, "getCertificate", {
  enumerable: true,
  get: function () {
    return _getCertificate().default;
  }
});
Object.defineProperty(exports, "getRootDir", {
  enumerable: true,
  get: function () {
    return _getRootDir().default;
  }
});
Object.defineProperty(exports, "isDirectoryInside", {
  enumerable: true,
  get: function () {
    return _isDirectoryInside().default;
  }
});
Object.defineProperty(exports, "isURL", {
  enumerable: true,
  get: function () {
    return _isUrl().default;
  }
});
Object.defineProperty(exports, "objectHash", {
  enumerable: true,
  get: function () {
    return _objectHash().default;
  }
});
Object.defineProperty(exports, "prettifyTime", {
  enumerable: true,
  get: function () {
    return _prettifyTime().default;
  }
});
Object.defineProperty(exports, "prettyDiagnostic", {
  enumerable: true,
  get: function () {
    return _prettyDiagnostic().default;
  }
});
Object.defineProperty(exports, "PromiseQueue", {
  enumerable: true,
  get: function () {
    return _PromiseQueue().default;
  }
});
Object.defineProperty(exports, "validateSchema", {
  enumerable: true,
  get: function () {
    return _schema().default;
  }
});
Object.defineProperty(exports, "fuzzySearch", {
  enumerable: true,
  get: function () {
    return _schema().fuzzySearch;
  }
});
Object.defineProperty(exports, "TapStream", {
  enumerable: true,
  get: function () {
    return _TapStream().default;
  }
});
Object.defineProperty(exports, "urlJoin", {
  enumerable: true,
  get: function () {
    return _urlJoin().default;
  }
});
Object.defineProperty(exports, "relativeUrl", {
  enumerable: true,
  get: function () {
    return _relativeUrl().default;
  }
});
Object.defineProperty(exports, "createDependencyLocation", {
  enumerable: true,
  get: function () {
    return _dependencyLocation().default;
  }
});
Object.defineProperty(exports, "debounce", {
  enumerable: true,
  get: function () {
    return _debounce().default;
  }
});
Object.defineProperty(exports, "throttle", {
  enumerable: true,
  get: function () {
    return _throttle().default;
  }
});
Object.defineProperty(exports, "openInBrowser", {
  enumerable: true,
  get: function () {
    return _openInBrowser().default;
  }
});
Object.defineProperty(exports, "findAlternativeNodeModules", {
  enumerable: true,
  get: function () {
    return _alternatives().findAlternativeNodeModules;
  }
});
Object.defineProperty(exports, "findAlternativeFiles", {
  enumerable: true,
  get: function () {
    return _alternatives().findAlternativeFiles;
  }
});
Object.defineProperty(exports, "blobToBuffer", {
  enumerable: true,
  get: function () {
    return _blob().blobToBuffer;
  }
});
Object.defineProperty(exports, "blobToString", {
  enumerable: true,
  get: function () {
    return _blob().blobToString;
  }
});
Object.defineProperty(exports, "unique", {
  enumerable: true,
  get: function () {
    return _collection().unique;
  }
});
Object.defineProperty(exports, "objectSortedEntries", {
  enumerable: true,
  get: function () {
    return _collection().objectSortedEntries;
  }
});
Object.defineProperty(exports, "objectSortedEntriesDeep", {
  enumerable: true,
  get: function () {
    return _collection().objectSortedEntriesDeep;
  }
});
Object.defineProperty(exports, "setDifference", {
  enumerable: true,
  get: function () {
    return _collection().setDifference;
  }
});
Object.defineProperty(exports, "resolveConfig", {
  enumerable: true,
  get: function () {
    return _config().resolveConfig;
  }
});
Object.defineProperty(exports, "resolveConfigSync", {
  enumerable: true,
  get: function () {
    return _config().resolveConfigSync;
  }
});
Object.defineProperty(exports, "loadConfig", {
  enumerable: true,
  get: function () {
    return _config().loadConfig;
  }
});
Object.defineProperty(exports, "DefaultMap", {
  enumerable: true,
  get: function () {
    return _DefaultMap().DefaultMap;
  }
});
Object.defineProperty(exports, "DefaultWeakMap", {
  enumerable: true,
  get: function () {
    return _DefaultMap().DefaultWeakMap;
  }
});
Object.defineProperty(exports, "makeDeferredWithPromise", {
  enumerable: true,
  get: function () {
    return _Deferred().makeDeferredWithPromise;
  }
});
Object.defineProperty(exports, "isGlob", {
  enumerable: true,
  get: function () {
    return _glob().isGlob;
  }
});
Object.defineProperty(exports, "isGlobMatch", {
  enumerable: true,
  get: function () {
    return _glob().isGlobMatch;
  }
});
Object.defineProperty(exports, "globSync", {
  enumerable: true,
  get: function () {
    return _glob().globSync;
  }
});
Object.defineProperty(exports, "glob", {
  enumerable: true,
  get: function () {
    return _glob().glob;
  }
});
Object.defineProperty(exports, "hashStream", {
  enumerable: true,
  get: function () {
    return _hash().hashStream;
  }
});
Object.defineProperty(exports, "hashObject", {
  enumerable: true,
  get: function () {
    return _hash().hashObject;
  }
});
Object.defineProperty(exports, "hashFile", {
  enumerable: true,
  get: function () {
    return _hash().hashFile;
  }
});
Object.defineProperty(exports, "createHTTPServer", {
  enumerable: true,
  get: function () {
    return _httpServer().createHTTPServer;
  }
});
Object.defineProperty(exports, "normalizeSeparators", {
  enumerable: true,
  get: function () {
    return _path().normalizeSeparators;
  }
});
Object.defineProperty(exports, "normalizePath", {
  enumerable: true,
  get: function () {
    return _path().normalizePath;
  }
});
Object.defineProperty(exports, "relativePath", {
  enumerable: true,
  get: function () {
    return _path().relativePath;
  }
});
Object.defineProperty(exports, "replaceURLReferences", {
  enumerable: true,
  get: function () {
    return _replaceBundleReferences().replaceURLReferences;
  }
});
Object.defineProperty(exports, "replaceInlineReferences", {
  enumerable: true,
  get: function () {
    return _replaceBundleReferences().replaceInlineReferences;
  }
});
Object.defineProperty(exports, "measureStreamLength", {
  enumerable: true,
  get: function () {
    return _stream().measureStreamLength;
  }
});
Object.defineProperty(exports, "readableFromStringOrBuffer", {
  enumerable: true,
  get: function () {
    return _stream().readableFromStringOrBuffer;
  }
});
Object.defineProperty(exports, "bufferStream", {
  enumerable: true,
  get: function () {
    return _stream().bufferStream;
  }
});
Object.defineProperty(exports, "blobToStream", {
  enumerable: true,
  get: function () {
    return _stream().blobToStream;
  }
});
Object.defineProperty(exports, "streamFromPromise", {
  enumerable: true,
  get: function () {
    return _stream().streamFromPromise;
  }
});
Object.defineProperty(exports, "fallbackStream", {
  enumerable: true,
  get: function () {
    return _stream().fallbackStream;
  }
});
Object.defineProperty(exports, "relativeBundlePath", {
  enumerable: true,
  get: function () {
    return _relativeBundlePath().relativeBundlePath;
  }
});
Object.defineProperty(exports, "ansiHtml", {
  enumerable: true,
  get: function () {
    return _ansiHtml().ansiHtml;
  }
});
Object.defineProperty(exports, "escapeHTML", {
  enumerable: true,
  get: function () {
    return _escapeHtml().escapeHTML;
  }
});
Object.defineProperty(exports, "SOURCEMAP_RE", {
  enumerable: true,
  get: function () {
    return _sourcemap().SOURCEMAP_RE;
  }
});
Object.defineProperty(exports, "SOURCEMAP_EXTENSIONS", {
  enumerable: true,
  get: function () {
    return _sourcemap().SOURCEMAP_EXTENSIONS;
  }
});
Object.defineProperty(exports, "matchSourceMappingURL", {
  enumerable: true,
  get: function () {
    return _sourcemap().matchSourceMappingURL;
  }
});
Object.defineProperty(exports, "loadSourceMapUrl", {
  enumerable: true,
  get: function () {
    return _sourcemap().loadSourceMapUrl;
  }
});
Object.defineProperty(exports, "loadSourceMap", {
  enumerable: true,
  get: function () {
    return _sourcemap().loadSourceMap;
  }
});
Object.defineProperty(exports, "remapSourceLocation", {
  enumerable: true,
  get: function () {
    return _sourcemap().remapSourceLocation;
  }
});

function _countLines() {
  const data = _interopRequireDefault(require("./countLines"));

  _countLines = function () {
    return data;
  };

  return data;
}

function _generateBuildMetrics() {
  const data = _interopRequireDefault(require("./generateBuildMetrics"));

  _generateBuildMetrics = function () {
    return data;
  };

  return data;
}

function _generateCertificate() {
  const data = _interopRequireDefault(require("./generateCertificate"));

  _generateCertificate = function () {
    return data;
  };

  return data;
}

function _getCertificate() {
  const data = _interopRequireDefault(require("./getCertificate"));

  _getCertificate = function () {
    return data;
  };

  return data;
}

function _getRootDir() {
  const data = _interopRequireDefault(require("./getRootDir"));

  _getRootDir = function () {
    return data;
  };

  return data;
}

function _isDirectoryInside() {
  const data = _interopRequireDefault(require("./isDirectoryInside"));

  _isDirectoryInside = function () {
    return data;
  };

  return data;
}

function _isUrl() {
  const data = _interopRequireDefault(require("./is-url"));

  _isUrl = function () {
    return data;
  };

  return data;
}

function _objectHash() {
  const data = _interopRequireDefault(require("./objectHash"));

  _objectHash = function () {
    return data;
  };

  return data;
}

function _prettifyTime() {
  const data = _interopRequireDefault(require("./prettifyTime"));

  _prettifyTime = function () {
    return data;
  };

  return data;
}

function _prettyDiagnostic() {
  const data = _interopRequireDefault(require("./prettyDiagnostic"));

  _prettyDiagnostic = function () {
    return data;
  };

  return data;
}

function _PromiseQueue() {
  const data = _interopRequireDefault(require("./PromiseQueue"));

  _PromiseQueue = function () {
    return data;
  };

  return data;
}

function _schema() {
  const data = _interopRequireWildcard(require("./schema"));

  _schema = function () {
    return data;
  };

  return data;
}

function _TapStream() {
  const data = _interopRequireDefault(require("./TapStream"));

  _TapStream = function () {
    return data;
  };

  return data;
}

function _urlJoin() {
  const data = _interopRequireDefault(require("./urlJoin"));

  _urlJoin = function () {
    return data;
  };

  return data;
}

function _relativeUrl() {
  const data = _interopRequireDefault(require("./relativeUrl"));

  _relativeUrl = function () {
    return data;
  };

  return data;
}

function _dependencyLocation() {
  const data = _interopRequireDefault(require("./dependency-location"));

  _dependencyLocation = function () {
    return data;
  };

  return data;
}

function _debounce() {
  const data = _interopRequireDefault(require("./debounce"));

  _debounce = function () {
    return data;
  };

  return data;
}

function _throttle() {
  const data = _interopRequireDefault(require("./throttle"));

  _throttle = function () {
    return data;
  };

  return data;
}

function _openInBrowser() {
  const data = _interopRequireDefault(require("./openInBrowser"));

  _openInBrowser = function () {
    return data;
  };

  return data;
}

function _alternatives() {
  const data = require("./alternatives");

  _alternatives = function () {
    return data;
  };

  return data;
}

function _blob() {
  const data = require("./blob");

  _blob = function () {
    return data;
  };

  return data;
}

function _collection() {
  const data = require("./collection");

  _collection = function () {
    return data;
  };

  return data;
}

function _config() {
  const data = require("./config");

  _config = function () {
    return data;
  };

  return data;
}

function _DefaultMap() {
  const data = require("./DefaultMap");

  _DefaultMap = function () {
    return data;
  };

  return data;
}

function _Deferred() {
  const data = require("./Deferred");

  _Deferred = function () {
    return data;
  };

  return data;
}

function _glob() {
  const data = require("./glob");

  _glob = function () {
    return data;
  };

  return data;
}

function _hash() {
  const data = require("./hash");

  _hash = function () {
    return data;
  };

  return data;
}

function _httpServer() {
  const data = require("./http-server");

  _httpServer = function () {
    return data;
  };

  return data;
}

function _path() {
  const data = require("./path");

  _path = function () {
    return data;
  };

  return data;
}

function _replaceBundleReferences() {
  const data = require("./replaceBundleReferences");

  _replaceBundleReferences = function () {
    return data;
  };

  return data;
}

function _stream() {
  const data = require("./stream");

  _stream = function () {
    return data;
  };

  return data;
}

function _relativeBundlePath() {
  const data = require("./relativeBundlePath");

  _relativeBundlePath = function () {
    return data;
  };

  return data;
}

function _ansiHtml() {
  const data = require("./ansi-html");

  _ansiHtml = function () {
    return data;
  };

  return data;
}

function _escapeHtml() {
  const data = require("./escape-html");

  _escapeHtml = function () {
    return data;
  };

  return data;
}

function _sourcemap() {
  const data = require("./sourcemap");

  _sourcemap = function () {
    return data;
  };

  return data;
}

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
