// @parcel/utils/lib/prettyDiagnostic.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = prettyDiagnostic;

function _codeframe() {
  const data = _interopRequireDefault(require("@parcel/codeframe"));

  _codeframe = function () {
    return data;
  };

  return data;
}

function _markdownAnsi() {
  const data = _interopRequireDefault(require("@parcel/markdown-ansi"));

  _markdownAnsi = function () {
    return data;
  };

  return data;
}

function _chalk() {
  const data = _interopRequireDefault(require("chalk"));

  _chalk = function () {
    return data;
  };

  return data;
}

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

function _nullthrows() {
  const data = _interopRequireDefault(require("nullthrows"));

  _nullthrows = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

async function prettyDiagnostic(diagnostic, options, terminalWidth) {
  let {
    origin,
    message,
    stack,
    codeFrame,
    hints,
    filePath,
    language,
    skipFormatting
  } = diagnostic;

  if (filePath != null && options && !_path().default.isAbsolute(filePath)) {
    filePath = _path().default.join(options.projectRoot, filePath);
  }

  let result = {
    message: (0, _markdownAnsi().default)(`**${origin !== null && origin !== void 0 ? origin : 'unknown'}**: `) + (skipFormatting ? message : (0, _markdownAnsi().default)(message)),
    stack: '',
    codeframe: '',
    hints: []
  };

  if (codeFrame != null) {
    var _codeFrame$code;

    let highlights = codeFrame.codeHighlights;
    let code = (_codeFrame$code = codeFrame.code) !== null && _codeFrame$code !== void 0 ? _codeFrame$code : options && (await options.inputFS.readFile((0, _nullthrows().default)(filePath), 'utf8'));
    let formattedCodeFrame = '';

    if (code != null) {
      formattedCodeFrame = (0, _codeframe().default)(code, highlights, {
        useColor: true,
        syntaxHighlighting: true,
        language: // $FlowFixMe sketchy null checks do not matter here...
        language || (filePath ? _path().default.extname(filePath).substr(1) : undefined),
        terminalWidth
      });
    }

    result.codeframe += typeof filePath !== 'string' ? '' : _chalk().default.underline(`${filePath}:${highlights[0].start.line}:${highlights[0].start.column}\n`);
    result.codeframe += formattedCodeFrame;
  } else if (typeof filePath === 'string') {
    result.codeframe += _chalk().default.underline(filePath);
  }

  if (stack != null) {
    result.stack = stack;
  } else if (filePath != null && result.codeframe == null) {
    result.stack = filePath;
  }

  if (Array.isArray(hints) && hints.length) {
    result.hints = hints.map(h => {
      return (0, _markdownAnsi().default)(h);
    });
  }

  return result;
}
