// @parcel/core/lib/requests/TargetRequest.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createTargetRequest;
exports.skipTarget = skipTarget;
exports.TargetResolver = void 0;

function _diagnostic() {
  const data = _interopRequireWildcard(require("@parcel/diagnostic"));

  _diagnostic = function () {
    return data;
  };

  return data;
}

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

function _utils() {
  const data = require("@parcel/utils");

  _utils = function () {
    return data;
  };

  return data;
}

function _Environment() {
  const data = require("../Environment");

  _Environment = function () {
    return data;
  };

  return data;
}

function _ParcelConfigRequest() {
  const data = _interopRequireWildcard(require("./ParcelConfigRequest"));

  _ParcelConfigRequest = function () {
    return data;
  };

  return data;
}

function _browserslist() {
  const data = _interopRequireDefault(require("browserslist"));

  _browserslist = function () {
    return data;
  };

  return data;
}

function _jsonSourceMap() {
  const data = _interopRequireDefault(require("json-source-map"));

  _jsonSourceMap = function () {
    return data;
  };

  return data;
}

function _assert() {
  const data = _interopRequireDefault(require("assert"));

  _assert = function () {
    return data;
  };

  return data;
}

function _nullthrows() {
  const data = _interopRequireDefault(require("nullthrows"));

  _nullthrows = function () {
    return data;
  };

  return data;
}

function _TargetDescriptor() {
  const data = require("../TargetDescriptor.schema");

  _TargetDescriptor = function () {
    return data;
  };

  return data;
}

function _Environment2() {
  const data = require("../public/Environment");

  _Environment2 = function () {
    return data;
  };

  return data;
}

function _utils2() {
  const data = require("../utils");

  _utils2 = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const DEFAULT_DIST_DIRNAME = 'dist';
const JS_RE = /\.[mc]?js$/;
const JS_EXTENSIONS = ['.js', '.mjs', '.cjs'];
const COMMON_TARGETS = {
  main: {
    match: JS_RE,
    extensions: JS_EXTENSIONS
  },
  module: {
    // module field is always ESM. Don't allow .cjs extension here.
    match: /\.m?js$/,
    extensions: ['.js', '.mjs']
  },
  browser: {
    match: JS_RE,
    extensions: JS_EXTENSIONS
  },
  types: {
    match: /\.d\.ts$/,
    extensions: ['.d.ts']
  }
};
const type = 'target_request';

function createTargetRequest(input) {
  return {
    id: `${type}:${(0, _utils().hashObject)(input)}`,
    type,
    run,
    input
  };
}

function skipTarget(targetName, exclusiveTarget, descriptorSource) {
  //  We skip targets if they have a descriptor.source and don't match the current exclusiveTarget
  //  They will be handled by a separate resolvePackageTargets call from their Entry point
  //  but with exclusiveTarget set.
  return exclusiveTarget == null ? descriptorSource != null : targetName !== exclusiveTarget;
}

async function run({
  input,
  api,
  options
}) {
  let targetResolver = new TargetResolver(api, (0, _utils2().optionsProxy)(options, api.invalidateOnOptionChange));
  let targets = await targetResolver.resolve(input.packagePath, input.target);
  let configResult = (0, _nullthrows().default)(await api.runRequest((0, _ParcelConfigRequest().default)()));
  let parcelConfig = (0, _ParcelConfigRequest().getCachedParcelConfig)(configResult, options); // Find named pipelines for each target.

  let pipelineNames = new Set(parcelConfig.getNamedPipelines());

  for (let target of targets) {
    if (pipelineNames.has(target.name)) {
      target.pipeline = target.name;
    }
  }

  return targets;
}

class TargetResolver {
  constructor(api, options) {
    _defineProperty(this, "fs", void 0);

    _defineProperty(this, "api", void 0);

    _defineProperty(this, "options", void 0);

    this.api = api;
    this.fs = options.inputFS;
    this.options = options;
  }

  async resolve(rootDir, exclusiveTarget) {
    let optionTargets = this.options.targets;

    if (exclusiveTarget != null && optionTargets == null) {
      optionTargets = [exclusiveTarget];
    }

    let packageTargets = await this.resolvePackageTargets(rootDir, exclusiveTarget);
    let targets;

    if (optionTargets) {
      if (Array.isArray(optionTargets)) {
        if (optionTargets.length === 0) {
          throw new (_diagnostic().default)({
            diagnostic: {
              message: `Targets option is an empty array`,
              origin: '@parcel/core'
            }
          });
        } // If an array of strings is passed, it's a filter on the resolved package
        // targets. Load them, and find the matching targets.


        targets = optionTargets.map(target => {
          let matchingTarget = packageTargets.get(target);

          if (!matchingTarget) {
            throw new (_diagnostic().default)({
              diagnostic: {
                message: (0, _diagnostic().md)`Could not find target with name "${target}"`,
                origin: '@parcel/core'
              }
            });
          }

          return matchingTarget;
        });
      } else {
        // Otherwise, it's an object map of target descriptors (similar to those
        // in package.json). Adapt them to native targets.
        targets = Object.entries(optionTargets).map(([name, _descriptor]) => {
          var _descriptor$publicUrl, _descriptor$outputFor;

          let {
            distDir,
            ...descriptor
          } = parseDescriptor(name, _descriptor, null, JSON.stringify({
            targets: optionTargets
          }, null, '\t'));

          if (distDir == null) {
            let optionTargetsString = JSON.stringify(optionTargets, null, '\t');
            throw new (_diagnostic().default)({
              diagnostic: {
                message: (0, _diagnostic().md)`Missing distDir for target "${name}"`,
                origin: '@parcel/core',
                codeFrame: {
                  code: optionTargetsString,
                  codeHighlights: (0, _diagnostic().generateJSONCodeHighlights)(optionTargetsString || '', [{
                    key: `/${name}`,
                    type: 'value'
                  }])
                }
              }
            });
          }

          let target = {
            name,
            distDir: _path().default.resolve(this.fs.cwd(), distDir),
            publicUrl: (_descriptor$publicUrl = descriptor.publicUrl) !== null && _descriptor$publicUrl !== void 0 ? _descriptor$publicUrl : this.options.defaultTargetOptions.publicUrl,
            env: (0, _Environment().createEnvironment)({
              engines: descriptor.engines,
              context: descriptor.context,
              isLibrary: descriptor.isLibrary,
              includeNodeModules: descriptor.includeNodeModules,
              outputFormat: (_descriptor$outputFor = descriptor.outputFormat) !== null && _descriptor$outputFor !== void 0 ? _descriptor$outputFor : this.options.defaultTargetOptions.outputFormat,
              shouldOptimize: this.options.defaultTargetOptions.shouldOptimize && descriptor.optimize !== false,
              shouldScopeHoist: this.options.defaultTargetOptions.shouldScopeHoist && descriptor.scopeHoist !== false,
              sourceMap: normalizeSourceMap(this.options, descriptor.sourceMap)
            })
          };

          if (descriptor.distEntry != null) {
            target.distEntry = descriptor.distEntry;
          }

          if (descriptor.source != null) {
            target.source = descriptor.source;
          }

          return target;
        }).filter(target => !skipTarget(target.name, exclusiveTarget, target.source));
      }

      let serve = this.options.serveOptions;

      if (serve) {
        // In serve mode, we only support a single browser target. If the user
        // provided more than one, or the matching target is not a browser, throw.
        if (targets.length > 1) {
          throw new (_diagnostic().default)({
            diagnostic: {
              message: `More than one target is not supported in serve mode`,
              origin: '@parcel/core'
            }
          });
        }

        if (!_Environment2().BROWSER_ENVS.has(targets[0].env.context)) {
          throw new (_diagnostic().default)({
            diagnostic: {
              message: `Only browser targets are supported in serve mode`,
              origin: '@parcel/core'
            }
          });
        }

        targets[0].distDir = serve.distDir;
      }
    } else {
      // Explicit targets were not provided. Either use a modern target for server
      // mode, or simply use the package.json targets.
      if (this.options.serveOptions) {
        var _this$options$default;

        // In serve mode, we only support a single browser target. Since the user
        // hasn't specified a target, use one targeting modern browsers for development
        targets = [{
          name: 'default',
          distDir: this.options.serveOptions.distDir,
          publicUrl: (_this$options$default = this.options.defaultTargetOptions.publicUrl) !== null && _this$options$default !== void 0 ? _this$options$default : '/',
          env: (0, _Environment().createEnvironment)({
            context: 'browser',
            engines: {},
            shouldOptimize: this.options.defaultTargetOptions.shouldOptimize,
            outputFormat: this.options.defaultTargetOptions.outputFormat,
            shouldScopeHoist: this.options.defaultTargetOptions.shouldScopeHoist,
            sourceMap: this.options.defaultTargetOptions.sourceMaps ? {} : undefined
          })
        }];
      } else {
        targets = Array.from(packageTargets.values()).filter(descriptor => {
          return !skipTarget(descriptor.name, exclusiveTarget, descriptor.source);
        });
      }
    }

    return targets;
  }

  async resolvePackageTargets(rootDir, exclusiveTarget) {
    var _ref2, _pkg$browser, _pkg$browser2;

    let rootFile = _path().default.join(rootDir, 'index');

    let conf = await (0, _utils().loadConfig)(this.fs, rootFile, ['package.json'], this.options.projectRoot); // Invalidate whenever a package.json file is added.

    this.api.invalidateOnFileCreate({
      fileName: 'package.json',
      aboveFilePath: rootFile
    });
    let pkg;
    let pkgContents;
    let pkgFilePath;
    let pkgDir;
    let pkgMap;

    if (conf) {
      pkg = conf.config;
      let pkgFile = conf.files[0];

      if (pkgFile == null) {
        throw new (_diagnostic().default)({
          diagnostic: {
            message: (0, _diagnostic().md)`Expected package.json file in ${rootDir}`,
            origin: '@parcel/core'
          }
        });
      }

      let _pkgFilePath = pkgFilePath = pkgFile.filePath; // For Flow


      pkgDir = _path().default.dirname(_pkgFilePath);
      pkgContents = await this.fs.readFile(_pkgFilePath, 'utf8');
      pkgMap = _jsonSourceMap().default.parse(pkgContents.replace(/\t/g, ' '));
      this.api.invalidateOnFileUpdate(_pkgFilePath);
      this.api.invalidateOnFileDelete(_pkgFilePath);
    } else {
      pkg = {};
      pkgDir = this.fs.cwd();
    }

    let pkgTargets = pkg.targets || {};
    let pkgEngines = parseEngines(pkg.engines, pkgFilePath, pkgContents, '/engines', 'Invalid engines in package.json') || {};

    if (pkgEngines.browsers == null) {
      var _ref, _this$options$env$BRO;

      let env = (_ref = (_this$options$env$BRO = this.options.env.BROWSERSLIST_ENV) !== null && _this$options$env$BRO !== void 0 ? _this$options$env$BRO : this.options.env.NODE_ENV) !== null && _ref !== void 0 ? _ref : this.options.mode;

      if (pkg.browserslist != null) {
        let pkgBrowserslist = pkg.browserslist;
        let browserslist = typeof pkgBrowserslist === 'object' && !Array.isArray(pkgBrowserslist) ? pkgBrowserslist[env] : pkgBrowserslist;
        pkgEngines = { ...pkgEngines,
          browsers: browserslist
        };
      } else {
        let browserslistConfig = await (0, _utils().resolveConfig)(this.fs, _path().default.join(rootDir, 'index'), ['browserslist', '.browserslistrc'], this.options.projectRoot);
        this.api.invalidateOnFileCreate({
          fileName: 'browserslist',
          aboveFilePath: rootFile
        });
        this.api.invalidateOnFileCreate({
          fileName: '.browserslistrc',
          aboveFilePath: rootFile
        });

        if (browserslistConfig != null) {
          let contents = await this.fs.readFile(browserslistConfig, 'utf8');

          let config = _browserslist().default.parseConfig(contents);

          let browserslistBrowsers = config[env] || config.defaults;

          if ((browserslistBrowsers === null || browserslistBrowsers === void 0 ? void 0 : browserslistBrowsers.length) > 0) {
            pkgEngines = { ...pkgEngines,
              browsers: browserslistBrowsers
            };
          } // Invalidate whenever browserslist config file or relevant environment variables change


          this.api.invalidateOnFileUpdate(browserslistConfig);
          this.api.invalidateOnFileDelete(browserslistConfig);
          this.api.invalidateOnEnvChange('BROWSERSLIST_ENV');
          this.api.invalidateOnEnvChange('NODE_ENV');
        }
      }
    }

    let targets = new Map();
    let node = pkgEngines.node;
    let browsers = pkgEngines.browsers; // If there is a separate `browser` target, or an `engines.node` field but no browser targets, then
    // the `main` and `module` targets refer to node, otherwise browser.

    let mainContext = ((_ref2 = (_pkg$browser = pkg.browser) !== null && _pkg$browser !== void 0 ? _pkg$browser : pkgTargets.browser) !== null && _ref2 !== void 0 ? _ref2 : node != null && !browsers) ? 'node' : 'browser';
    let moduleContext = ((_pkg$browser2 = pkg.browser) !== null && _pkg$browser2 !== void 0 ? _pkg$browser2 : pkgTargets.browser) ? 'browser' : mainContext;
    let defaultEngines = this.options.defaultTargetOptions.engines;
    let context = (browsers !== null && browsers !== void 0 ? browsers : !node) ? 'browser' : 'node';

    if (context === 'browser' && pkgEngines.browsers == null && (defaultEngines === null || defaultEngines === void 0 ? void 0 : defaultEngines.browsers) != null) {
      pkgEngines = { ...pkgEngines,
        browsers: defaultEngines.browsers
      };
    } else if (context === 'node' && pkgEngines.node == null && (defaultEngines === null || defaultEngines === void 0 ? void 0 : defaultEngines.node) != null) {
      pkgEngines = { ...pkgEngines,
        node: defaultEngines.node
      };
    }

    for (let targetName in COMMON_TARGETS) {
      let _targetDist;

      let pointer;

      if (targetName === 'browser' && pkg[targetName] != null && typeof pkg[targetName] === 'object') {
        // The `browser` field can be a file path or an alias map.
        _targetDist = pkg[targetName][pkg.name];
        pointer = `/${targetName}/${pkg.name}`;
      } else {
        _targetDist = pkg[targetName];
        pointer = `/${targetName}`;
      } // For Flow


      let targetDist = _targetDist;

      if (typeof targetDist === 'string' || pkgTargets[targetName]) {
        var _pkgTargets$targetNam, _ref3, _ref4, _descriptor$outputFor2, _descriptor$publicUrl2, _descriptor$engines, _descriptor$context, _descriptor$includeNo;

        let distDir;
        let distEntry;
        let loc;
        (0, _assert().default)(pkgMap != null);

        let _descriptor = (_pkgTargets$targetNam = pkgTargets[targetName]) !== null && _pkgTargets$targetNam !== void 0 ? _pkgTargets$targetNam : {};

        if (typeof targetDist === 'string') {
          distDir = _path().default.resolve(pkgDir, _path().default.dirname(targetDist));
          distEntry = _path().default.basename(targetDist);
          loc = {
            filePath: (0, _nullthrows().default)(pkgFilePath),
            ...(0, _diagnostic().getJSONSourceLocation)(pkgMap.pointers[pointer], 'value')
          };
        } else {
          var _this$options$default2;

          distDir = (_this$options$default2 = this.options.defaultTargetOptions.distDir) !== null && _this$options$default2 !== void 0 ? _this$options$default2 : _path().default.join(pkgDir, DEFAULT_DIST_DIRNAME, targetName);
        }

        if (_descriptor == false) {
          continue;
        }

        let descriptor = parseCommonTargetDescriptor(targetName, _descriptor, pkgFilePath, pkgContents);

        if (skipTarget(targetName, exclusiveTarget, descriptor.source)) {
          continue;
        }

        if (distEntry != null && !COMMON_TARGETS[targetName].match.test(distEntry)) {
          var _pkgFilePath2;

          let contents = typeof pkgContents === 'string' ? pkgContents : // $FlowFixMe
          JSON.stringify(pkgContents, null, '\t'); // $FlowFixMe

          let listFormat = new Intl.ListFormat('en-US', {
            type: 'disjunction'
          });
          let extensions = listFormat.format(COMMON_TARGETS[targetName].extensions);

          let ext = _path().default.extname(distEntry);

          throw new (_diagnostic().default)({
            diagnostic: {
              message: (0, _diagnostic().md)`Unexpected output file type ${ext} in target "${targetName}"`,
              origin: '@parcel/core',
              language: 'json',
              filePath: (_pkgFilePath2 = pkgFilePath) !== null && _pkgFilePath2 !== void 0 ? _pkgFilePath2 : undefined,
              codeFrame: {
                code: contents,
                codeHighlights: (0, _diagnostic().generateJSONCodeHighlights)(contents, [{
                  key: pointer,
                  type: 'value',
                  message: `File extension must be ${extensions}`
                }])
              },
              hints: [`The "${targetName}" field is meant for libraries. If you meant to output a ${ext} file, either remove the "${targetName}" field or choose a different target name.`]
            }
          });
        }

        if (descriptor.outputFormat === 'global') {
          var _pkgFilePath3;

          let contents = typeof pkgContents === 'string' ? pkgContents : // $FlowFixMe
          JSON.stringify(pkgContents, null, '\t');
          throw new (_diagnostic().default)({
            diagnostic: {
              message: (0, _diagnostic().md)`The "global" output format is not supported in the "${targetName}" target.`,
              origin: '@parcel/core',
              language: 'json',
              filePath: (_pkgFilePath3 = pkgFilePath) !== null && _pkgFilePath3 !== void 0 ? _pkgFilePath3 : undefined,
              codeFrame: {
                code: contents,
                codeHighlights: (0, _diagnostic().generateJSONCodeHighlights)(contents, [{
                  key: `/targets/${targetName}/outputFormat`,
                  type: 'value'
                }])
              },
              hints: [`The "${targetName}" field is meant for libraries. The outputFormat must be either "commonjs" or "esmodule". Either change or remove the declared outputFormat.`]
            }
          });
        }

        let inferredOutputFormat = this.inferOutputFormat(distEntry, descriptor, targetName, pkg, pkgFilePath, pkgContents);
        let outputFormat = (_ref3 = (_ref4 = (_descriptor$outputFor2 = descriptor.outputFormat) !== null && _descriptor$outputFor2 !== void 0 ? _descriptor$outputFor2 : this.options.defaultTargetOptions.outputFormat) !== null && _ref4 !== void 0 ? _ref4 : inferredOutputFormat) !== null && _ref3 !== void 0 ? _ref3 : targetName === 'module' ? 'esmodule' : 'commonjs';

        if (targetName === 'main' && outputFormat === 'esmodule' && inferredOutputFormat !== 'esmodule') {
          var _pkgFilePath4;

          let contents = typeof pkgContents === 'string' ? pkgContents : // $FlowFixMe
          JSON.stringify(pkgContents, null, '\t');
          throw new (_diagnostic().default)({
            diagnostic: {
              // prettier-ignore
              message: (0, _diagnostic().md)`Output format "esmodule" cannot be used in the "main" target without a .mjs extension or "type": "module" field.`,
              origin: '@parcel/core',
              language: 'json',
              filePath: (_pkgFilePath4 = pkgFilePath) !== null && _pkgFilePath4 !== void 0 ? _pkgFilePath4 : undefined,
              codeFrame: {
                code: contents,
                codeHighlights: (0, _diagnostic().generateJSONCodeHighlights)(contents, [{
                  key: `/targets/${targetName}/outputFormat`,
                  type: 'value',
                  message: 'Declared output format defined here'
                }, {
                  key: '/main',
                  type: 'value',
                  message: 'Inferred output format defined here'
                }])
              },
              hints: [`Either change the output file extension to .mjs, add "type": "module" to package.json, or remove the declared outputFormat.`]
            }
          });
        }

        targets.set(targetName, {
          name: targetName,
          distDir,
          distEntry,
          publicUrl: (_descriptor$publicUrl2 = descriptor.publicUrl) !== null && _descriptor$publicUrl2 !== void 0 ? _descriptor$publicUrl2 : this.options.defaultTargetOptions.publicUrl,
          env: (0, _Environment().createEnvironment)({
            engines: (_descriptor$engines = descriptor.engines) !== null && _descriptor$engines !== void 0 ? _descriptor$engines : pkgEngines,
            context: (_descriptor$context = descriptor.context) !== null && _descriptor$context !== void 0 ? _descriptor$context : targetName === 'browser' ? 'browser' : outputFormat === 'esmodule' ? moduleContext : mainContext,
            includeNodeModules: (_descriptor$includeNo = descriptor.includeNodeModules) !== null && _descriptor$includeNo !== void 0 ? _descriptor$includeNo : false,
            outputFormat,
            isLibrary: true,
            shouldOptimize: this.options.defaultTargetOptions.shouldOptimize && descriptor.optimize !== false,
            shouldScopeHoist: this.options.defaultTargetOptions.shouldScopeHoist && descriptor.scopeHoist !== false,
            sourceMap: normalizeSourceMap(this.options, descriptor.sourceMap)
          }),
          loc
        });
      }
    }

    let customTargets = Object.keys(pkgTargets).filter(targetName => !COMMON_TARGETS[targetName]); // Custom targets

    for (let targetName of customTargets) {
      let distPath = pkg[targetName];
      let distDir;
      let distEntry;
      let loc;

      if (distPath == null) {
        var _this$options$default3;

        distDir = (_this$options$default3 = this.options.defaultTargetOptions.distDir) !== null && _this$options$default3 !== void 0 ? _this$options$default3 : _path().default.join(pkgDir, DEFAULT_DIST_DIRNAME);

        if (customTargets.length >= 2) {
          distDir = _path().default.join(distDir, targetName);
        }
      } else {
        if (typeof distPath !== 'string') {
          var _pkgFilePath5;

          let contents = typeof pkgContents === 'string' ? pkgContents : // $FlowFixMe
          JSON.stringify(pkgContents, null, '\t');
          throw new (_diagnostic().default)({
            diagnostic: {
              message: (0, _diagnostic().md)`Invalid distPath for target "${targetName}"`,
              origin: '@parcel/core',
              language: 'json',
              filePath: (_pkgFilePath5 = pkgFilePath) !== null && _pkgFilePath5 !== void 0 ? _pkgFilePath5 : undefined,
              codeFrame: {
                code: contents,
                codeHighlights: (0, _diagnostic().generateJSONCodeHighlights)(contents, [{
                  key: `/${targetName}`,
                  type: 'value',
                  message: 'Expected type string'
                }])
              }
            }
          });
        }

        distDir = _path().default.resolve(pkgDir, _path().default.dirname(distPath));
        distEntry = _path().default.basename(distPath);
        (0, _assert().default)(typeof pkgFilePath === 'string');
        (0, _assert().default)(pkgMap != null);
        loc = {
          filePath: pkgFilePath,
          ...(0, _diagnostic().getJSONSourceLocation)(pkgMap.pointers[`/${targetName}`], 'value')
        };
      }

      if (targetName in pkgTargets) {
        var _descriptor$publicUrl3, _descriptor$engines2, _ref5, _ref6, _descriptor$outputFor3;

        let descriptor = parsePackageDescriptor(targetName, pkgTargets[targetName], pkgFilePath, pkgContents);

        let pkgDir = _path().default.dirname((0, _nullthrows().default)(pkgFilePath));

        if (skipTarget(targetName, exclusiveTarget, descriptor.source)) {
          continue;
        }

        let inferredOutputFormat = this.inferOutputFormat(distEntry, descriptor, targetName, pkg, pkgFilePath, pkgContents);
        targets.set(targetName, {
          name: targetName,
          distDir: descriptor.distDir != null ? _path().default.resolve(pkgDir, descriptor.distDir) : distDir,
          distEntry,
          publicUrl: (_descriptor$publicUrl3 = descriptor.publicUrl) !== null && _descriptor$publicUrl3 !== void 0 ? _descriptor$publicUrl3 : this.options.defaultTargetOptions.publicUrl,
          env: (0, _Environment().createEnvironment)({
            engines: (_descriptor$engines2 = descriptor.engines) !== null && _descriptor$engines2 !== void 0 ? _descriptor$engines2 : pkgEngines,
            context: descriptor.context,
            includeNodeModules: descriptor.includeNodeModules,
            outputFormat: (_ref5 = (_ref6 = (_descriptor$outputFor3 = descriptor.outputFormat) !== null && _descriptor$outputFor3 !== void 0 ? _descriptor$outputFor3 : this.options.defaultTargetOptions.outputFormat) !== null && _ref6 !== void 0 ? _ref6 : inferredOutputFormat) !== null && _ref5 !== void 0 ? _ref5 : undefined,
            isLibrary: descriptor.isLibrary,
            shouldOptimize: this.options.defaultTargetOptions.shouldOptimize && descriptor.optimize !== false,
            shouldScopeHoist: this.options.defaultTargetOptions.shouldScopeHoist && descriptor.scopeHoist !== false,
            sourceMap: normalizeSourceMap(this.options, descriptor.sourceMap)
          }),
          loc
        });
      }
    } // If no explicit targets were defined, add a default.


    if (targets.size === 0) {
      var _this$options$default4;

      targets.set('default', {
        name: 'default',
        distDir: (_this$options$default4 = this.options.defaultTargetOptions.distDir) !== null && _this$options$default4 !== void 0 ? _this$options$default4 : _path().default.join(pkgDir, DEFAULT_DIST_DIRNAME),
        publicUrl: this.options.defaultTargetOptions.publicUrl,
        env: (0, _Environment().createEnvironment)({
          engines: pkgEngines,
          context,
          outputFormat: this.options.defaultTargetOptions.outputFormat,
          shouldOptimize: this.options.defaultTargetOptions.shouldOptimize,
          shouldScopeHoist: this.options.defaultTargetOptions.shouldScopeHoist,
          sourceMap: this.options.defaultTargetOptions.sourceMaps ? {} : undefined
        })
      });
    }

    assertNoDuplicateTargets(targets, pkgFilePath, pkgContents);
    return targets;
  }

  inferOutputFormat(distEntry, descriptor, targetName, pkg, pkgFilePath, pkgContents) {
    // Infer the outputFormat based on package.json properties.
    // If the extension is .mjs it's always a module.
    // If the extension is .cjs, it's always commonjs.
    // If the "type" field is set to "module" and the extension is .js, it's a module.
    let ext = distEntry != null ? _path().default.extname(distEntry) : null;
    let inferredOutputFormat, inferredOutputFormatField;

    switch (ext) {
      case '.mjs':
        inferredOutputFormat = 'esmodule';
        inferredOutputFormatField = `/${targetName}`;
        break;

      case '.cjs':
        inferredOutputFormat = 'commonjs';
        inferredOutputFormatField = `/${targetName}`;
        break;

      case '.js':
        if (pkg.type === 'module') {
          inferredOutputFormat = 'esmodule';
          inferredOutputFormatField = '/type';
        }

        break;
    }

    if (descriptor.outputFormat && inferredOutputFormat && descriptor.outputFormat !== inferredOutputFormat) {
      let contents = typeof pkgContents === 'string' ? pkgContents : // $FlowFixMe
      JSON.stringify(pkgContents, null, '\t');
      let expectedExtensions;

      switch (descriptor.outputFormat) {
        case 'esmodule':
          expectedExtensions = ['.mjs', '.js'];
          break;

        case 'commonjs':
          expectedExtensions = ['.cjs', '.js'];
          break;

        case 'global':
          expectedExtensions = ['.js'];
          break;
      } // $FlowFixMe


      let listFormat = new Intl.ListFormat('en-US', {
        type: 'disjunction'
      });
      throw new (_diagnostic().default)({
        diagnostic: {
          message: (0, _diagnostic().md)`Declared output format "${descriptor.outputFormat}" does not match expected output format "${inferredOutputFormat}".`,
          origin: '@parcel/core',
          language: 'json',
          filePath: pkgFilePath !== null && pkgFilePath !== void 0 ? pkgFilePath : undefined,
          codeFrame: {
            code: contents,
            codeHighlights: (0, _diagnostic().generateJSONCodeHighlights)(contents, [{
              key: `/targets/${targetName}/outputFormat`,
              type: 'value',
              message: 'Declared output format defined here'
            }, {
              key: (0, _nullthrows().default)(inferredOutputFormatField),
              type: 'value',
              message: 'Inferred output format defined here'
            }])
          },
          hints: [inferredOutputFormatField === '/type' ? 'Either remove the target\'s declared "outputFormat" or remove the "type" field.' : `Either remove the target's declared "outputFormat" or change the extension to ${listFormat.format(expectedExtensions)}.`]
        }
      });
    }

    return inferredOutputFormat;
  }

}

exports.TargetResolver = TargetResolver;

function parseEngines(engines, pkgPath, pkgContents, prependKey, message) {
  if (engines === undefined) {
    return engines;
  } else {
    _utils().validateSchema.diagnostic(_TargetDescriptor().ENGINES_SCHEMA, {
      data: engines,
      source: pkgContents,
      filePath: pkgPath,
      prependKey
    }, '@parcel/core', message); // $FlowFixMe we just verified this


    return engines;
  }
}

function parseDescriptor(targetName, descriptor, pkgPath, pkgContents) {
  _utils().validateSchema.diagnostic(_TargetDescriptor().DESCRIPTOR_SCHEMA, {
    data: descriptor,
    source: pkgContents,
    filePath: pkgPath,
    prependKey: `/targets/${targetName}`
  }, '@parcel/core', `Invalid target descriptor for target "${targetName}"`); // $FlowFixMe we just verified this


  return descriptor;
}

function parsePackageDescriptor(targetName, descriptor, pkgPath, pkgContents) {
  _utils().validateSchema.diagnostic(_TargetDescriptor().PACKAGE_DESCRIPTOR_SCHEMA, {
    data: descriptor,
    source: pkgContents,
    filePath: pkgPath,
    prependKey: `/targets/${targetName}`
  }, '@parcel/core', `Invalid target descriptor for target "${targetName}"`); // $FlowFixMe we just verified this


  return descriptor;
}

function parseCommonTargetDescriptor(targetName, descriptor, pkgPath, pkgContents) {
  _utils().validateSchema.diagnostic(_TargetDescriptor().COMMON_TARGET_DESCRIPTOR_SCHEMA, {
    data: descriptor,
    source: pkgContents,
    filePath: pkgPath,
    prependKey: `/targets/${targetName}`
  }, '@parcel/core', `Invalid target descriptor for target "${targetName}"`); // $FlowFixMe we just verified this


  return descriptor;
}

function assertNoDuplicateTargets(targets, pkgFilePath, pkgContents) {
  // Detect duplicate targets by destination path and provide a nice error.
  // Without this, an assertion is thrown much later after naming the bundles and finding duplicates.
  let targetsByPath = new Map();

  for (let target of targets.values()) {
    if (target.distEntry != null) {
      var _targetsByPath$get;

      let distPath = _path().default.join(target.distDir, target.distEntry);

      if (!targetsByPath.has(distPath)) {
        targetsByPath.set(distPath, []);
      }

      (_targetsByPath$get = targetsByPath.get(distPath)) === null || _targetsByPath$get === void 0 ? void 0 : _targetsByPath$get.push(target.name);
    }
  }

  let diagnostics = [];

  for (let [targetPath, targetNames] of targetsByPath) {
    if (targetNames.length > 1 && pkgContents != null && pkgFilePath != null) {
      diagnostics.push({
        message: (0, _diagnostic().md)`Multiple targets have the same destination path "${_path().default.relative(_path().default.dirname(pkgFilePath), targetPath)}"`,
        origin: '@parcel/core',
        language: 'json',
        filePath: pkgFilePath || undefined,
        codeFrame: {
          code: pkgContents,
          codeHighlights: (0, _diagnostic().generateJSONCodeHighlights)(pkgContents, targetNames.map(t => ({
            key: `/${t}`,
            type: 'value'
          })))
        }
      });
    }
  }

  if (diagnostics.length > 0) {
    // Only add hints to the last diagnostic so it isn't duplicated on each one
    diagnostics[diagnostics.length - 1].hints = ['Try removing the duplicate targets, or changing the destination paths.'];
    throw new (_diagnostic().default)({
      diagnostic: diagnostics
    });
  }
}

function normalizeSourceMap(options, sourceMap) {
  if (options.defaultTargetOptions.sourceMaps) {
    if (typeof sourceMap === 'boolean') {
      return sourceMap ? {} : undefined;
    } else {
      return sourceMap !== null && sourceMap !== void 0 ? sourceMap : {};
    }
  } else {
    return undefined;
  }
}
